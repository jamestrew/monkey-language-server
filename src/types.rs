#![allow(dead_code)]
use std::ops::{Deref, Range};

#[derive(PartialEq, Default)]
pub struct Spanned<T> {
    pub start: Position,
    pub end: Position,
    pub span: Range<usize>,
    data: T,
}

impl<T> Spanned<T> {
    pub fn new(start: Position, end: Position, span: Range<usize>, data: T) -> Self {
        Self {
            start,
            end,
            span,
            data,
        }
    }

    pub fn map<S>(&self, new_data: S) -> Spanned<S> {
        Spanned {
            start: self.start,
            end: self.end,
            span: self.span(),
            data: new_data,
        }
    }

    pub fn span(&self) -> Range<usize> {
        self.span.to_owned()
    }

    pub fn start(&self) -> usize {
        self.span.start
    }

    pub fn end(&self) -> usize {
        self.span.end
    }

    pub fn take(self) -> T {
        self.data
    }

    pub fn pos_rng_str(&self) -> String {
        format!("{:?}->{:?}", self.start, self.end)
}
}

impl<T> Deref for Spanned<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.data
    }
}

#[derive(Clone, Copy, PartialEq, Default)]
pub struct Position {
    row: usize,
    col: usize,
}

impl Position {
    pub fn new(row: usize, col: usize) -> Self {
        Self { row, col }
    }
}

impl std::fmt::Debug for Position {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({},{})", self.row, self.col)
    }
}
