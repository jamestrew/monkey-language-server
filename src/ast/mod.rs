mod expressions;
mod statements;

use std::fmt::Debug;

pub use expressions::*;
pub use statements::*;

use crate::diagnostics::SpannedError;

pub struct Program<'source> {
    pub nodes: Vec<Node<'source>>,
}

impl<'source> Program<'source> {
    pub fn collect_errors(&self) -> Vec<SpannedError> {
        let mut errors = Vec::new();

        for node in &self.nodes {
            match node {
                Node::Statement(stmt) => errors.extend(stmt.errors()),
                Node::Error(err) => errors.push(err.clone()),
            }
        }

        errors
    }
}

impl<'source> Debug for Program<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if f.alternate() {
            write!(f, "{:#?}", self.nodes)
        } else {
            write!(f, "{:?}", self.nodes)
        }
    }
}

trait NodeError {
    fn errors(&self) -> Vec<SpannedError>;
}

#[derive(PartialEq)]
pub enum Node<'source> {
    Statement(Statement<'source>),
    Error(SpannedError),
}

impl<'source> Debug for Node<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Statement(arg0) => write!(f, "{:#?}", arg0),
            Self::Error(arg0) => write!(f, "{:#?}", arg0),
        }
    }
}

impl<'source> NodeError for Node<'source> {
    fn errors(&self) -> Vec<SpannedError> {
        match &self {
            Node::Statement(stmt) => stmt.errors(),
            Node::Error(err) => vec![err.clone()],
        }
    }
}

impl<'source> From<Statement<'source>> for Node<'source> {
    fn from(stmt: Statement<'source>) -> Self {
        Node::Statement(stmt)
    }
}

impl<'source> From<Expression<'source>> for Node<'source> {
    fn from(expr: Expression<'source>) -> Self {
        Node::Statement(Statement::Expression(expr))
    }
}
