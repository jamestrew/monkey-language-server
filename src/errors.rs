use thiserror::Error;

use crate::types::Spanned;

pub type SpannedError = Spanned<MonkeyError>;

#[derive(Error, Debug, Clone, PartialEq)]
pub enum MonkeyError {
    #[error("Unexpected token '{0}'.")]
    UnexpectedToken(String),

    #[error("Unexpected eof.")]
    UnexpectedEof,

    #[error("Expected token '{0}' not found.")]
    ExpectedTokenNotFound(String),
}

impl std::fmt::Debug for SpannedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let err = match &**self {
            MonkeyError::UnexpectedToken(_) => format!("UnexpectedToken(\"{}\")", **self),
            MonkeyError::UnexpectedEof => format!("UnexpectedEof(\"{}\")", **self),
            MonkeyError::ExpectedTokenNotFound(_) => {
                format!("ExpectedTokenNotFound(\"{}\")", **self)
            }
        };
        write!(f, "Err({err}, {})", self.pos_rng_str())
    }
}
