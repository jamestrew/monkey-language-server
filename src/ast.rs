use std::fmt::Debug;

use crate::diagnostics::SpannedError;
use crate::lexer::{Token, TokenKind};

pub type StmtResult<'source> = Result<Statement<'source>, SpannedError>;
pub type ExprResult<'source> = Result<Expression<'source>, SpannedError>;
pub type BlockResult<'source> = Result<Block<'source>, SpannedError>;

pub struct Program<'source> {
    pub nodes: Vec<Node<'source>>,
}

impl<'source> Debug for Program<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if f.alternate() {
            write!(f, "{:#?}", self.nodes)
        } else {
            write!(f, "{:?}", self.nodes)
        }
    }
}

#[derive(PartialEq)]
pub enum Node<'source> {
    Statement(Statement<'source>),
    Error(SpannedError),
}

impl<'source> Debug for Node<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Statement(arg0) => write!(f, "{:#?}", arg0),
            Self::Error(arg0) => write!(f, "{:#?}", arg0),
        }
    }
}

impl<'source> From<Statement<'source>> for Node<'source> {
    fn from(stmt: Statement<'source>) -> Self {
        Node::Statement(stmt)
    }
}

impl<'source> From<Expression<'source>> for Statement<'source> {
    fn from(expr: Expression<'source>) -> Self {
        Statement::Expression(expr)
    }
}

impl<'source> From<Expression<'source>> for Node<'source> {
    fn from(expr: Expression<'source>) -> Self {
        Node::Statement(Statement::Expression(expr))
    }
}

#[derive(PartialEq)]
pub enum Statement<'source> {
    Let(Let<'source>),
    Return(Return<'source>),
    Block(Block<'source>),
    Expression(Expression<'source>),
}

impl<'source> Debug for Statement<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Let(arg0) => write!(f, "{:#?}", arg0),
            Self::Return(arg0) => write!(f, "{:#?}", arg0),
            Self::Block(arg0) => write!(f, "{:#?}", arg0),
            Self::Expression(arg0) => write!(f, "{:#?}", arg0),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Let<'source> {
    token: Token<'source>,
    name: Identifier<'source>,
    value: Expression<'source>,
}

impl<'source> Let<'source> {
    pub fn new(
        token: Token<'source>,
        name: Identifier<'source>,
        value: Expression<'source>,
    ) -> Self {
        Self { token, name, value }
    }
}

#[derive(Debug, PartialEq)]
pub struct Return<'source> {
    token: Token<'source>,
    value: Option<ExprResult<'source>>,
}

impl<'source> Return<'source> {
    pub fn new(token: Token<'source>, value: Option<ExprResult<'source>>) -> Self {
        Self { token, value }
    }
}

#[derive(Debug, PartialEq)]
pub struct Block<'source> {
    token: Token<'source>,
    statements: Vec<Node<'source>>,
}

impl<'source> Block<'source> {
    pub fn new(token: Token<'source>, statements: Vec<Node<'source>>) -> Self {
        Self { token, statements }
    }
}

macro_rules! stmt_impls {
    ($($stmt:tt),+) => {$(
        impl<'source> From<$stmt<'source>> for Node<'source> {
            fn from(stmt: $stmt<'source>) -> Self {
                let statement: Statement = stmt.into();
                statement.into()
            }
        }

        impl<'source> From<$stmt<'source>> for Statement<'source> {
            fn from(stmt: $stmt<'source>) -> Self {
                Statement::$stmt(stmt)
            }
        }
    )+}
}

stmt_impls!(Let, Return, Block);

#[derive(PartialEq)]
pub enum Expression<'source> {
    Identifier(Identifier<'source>),
    Primative(Primative<'source>),
    StringLiteral(StringLiteral<'source>),
    Prefix(Prefix<'source>),
    Infix(Infix<'source>),
    If(If<'source>),
    Function(Function<'source>),
    Call(Call<'source>),
    Array(Array<'source>),
    Hash(Hash<'source>),
    Index(Index<'source>),
}

impl<'source> Debug for Expression<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Identifier(arg0) => write!(f, "{:#?}", arg0),
            Self::Primative(arg0) => write!(f, "{:#?}", arg0),
            Self::StringLiteral(arg0) => write!(f, "{:#?}", arg0),
            Self::Prefix(arg0) => write!(f, "{:#?}", arg0),
            Self::Infix(arg0) => write!(f, "{:#?}", arg0),
            Self::If(arg0) => write!(f, "{:#?}", arg0),
            Self::Function(arg0) => write!(f, "{:#?}", arg0),
            Self::Call(arg0) => write!(f, "{:#?}", arg0),
            Self::Array(arg0) => write!(f, "{:#?}", arg0),
            Self::Hash(arg0) => write!(f, "{:#?}", arg0),
            Self::Index(arg0) => write!(f, "{:#?}", arg0),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Identifier<'source> {
    token: Token<'source>,
    name: &'source str,
}

impl<'source> From<Token<'source>> for Identifier<'source> {
    fn from(token: Token<'source>) -> Self {
        let name = token.slice;
        match token.kind {
            TokenKind::Identifier => Self { token, name },
            _ => unreachable!("Identifier expects Identifier token. got {:?}", token),
        }
    }
}

#[derive(Debug, PartialEq)]
pub enum Primative<'source> {
    Int { token: Token<'source>, value: i64 },
    Bool { token: Token<'source>, value: bool },
    Nil { token: Token<'source> },
}

impl<'source> From<Token<'source>> for Primative<'source> {
    fn from(token: Token<'source>) -> Self {
        match token.kind {
            TokenKind::Int => {
                let value: i64 = token.slice.parse().unwrap();
                Self::Int { token, value }
            }
            TokenKind::True => Self::Bool { token, value: true },
            TokenKind::False => Self::Bool {
                token,
                value: false,
            },
            TokenKind::Nil => Self::Nil { token },
            _ => unreachable!("Primative expects int/bool/nil. got {:?}", token),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct StringLiteral<'source> {
    token: Token<'source>,
    value: &'source str,
}

impl<'source> From<Token<'source>> for StringLiteral<'source> {
    fn from(token: Token<'source>) -> Self {
        let value = token.slice;
        match token.kind {
            TokenKind::Str => Self { token, value },
            _ => unreachable!("StringLiteral expects Str token. got {:?}", token),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Prefix<'source> {
    token: Token<'source>,
    right: Box<Expression<'source>>,
}

impl<'source> Prefix<'source> {
    pub fn new(token: Token<'source>, right: Expression<'source>) -> Self {
        Self {
            token,
            right: Box::new(right),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Infix<'source> {
    token: Token<'source>,
    left: Box<Expression<'source>>,
    right: Box<Expression<'source>>,
}

impl<'source> Infix<'source> {
    pub fn new(
        token: Token<'source>,
        left: Expression<'source>,
        right: Expression<'source>,
    ) -> Self {
        Self {
            token,
            left: Box::new(left),
            right: Box::new(right),
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct If<'source> {
    token: Token<'source>,
    condition: Result<Box<Expression<'source>>, SpannedError>,
    consequence: BlockResult<'source>,
    alternative: Result<Option<Block<'source>>, SpannedError>,
}

impl<'source> If<'source> {
    pub fn new(
        token: Token<'source>,
        condition: ExprResult<'source>,
        consequence: BlockResult<'source>,
        alternative: Result<Option<Block<'source>>, SpannedError>,
    ) -> Self {
        let condition = match condition {
            Ok(expr) => Ok(Box::new(expr)),
            Err(err) => Err(err),
        };

        Self {
            token,
            condition,
            consequence,
            alternative,
        }
    }
}

pub type VecExprResult<'source> = Result<Vec<ExprResult<'source>>, SpannedError>;

#[derive(Debug, PartialEq)]
pub struct Function<'source> {
    token: Token<'source>,
    params: VecExprResult<'source>,
    body: BlockResult<'source>,
}

impl<'source> Function<'source> {
    pub fn new(
        token: Token<'source>,
        params: VecExprResult<'source>,
        body: BlockResult<'source>,
    ) -> Self {
        Self {
            token,
            params,
            body,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Call<'source> {
    token: Token<'source>,
    func: Box<Expression<'source>>,
    args: VecExprResult<'source>,
}

impl<'source> Call<'source> {
    pub fn new(
        token: Token<'source>,
        func: Expression<'source>,
        args: VecExprResult<'source>,
    ) -> Self {
        Self {
            token,
            func: Box::new(func),
            args,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct Array<'source> {
    token: Token<'source>,
    elems: VecExprResult<'source>,
}

impl<'source> Array<'source> {
    pub fn new(token: Token<'source>, elems: VecExprResult<'source>) -> Self {
        Self { token, elems }
    }
}

pub type ExprPairs<'source> = (Expression<'source>, Expression<'source>);

#[derive(Debug, PartialEq)]
pub struct Hash<'source> {
    token: Token<'source>,
    kv_pairs: Result<Vec<Result<ExprPairs<'source>, SpannedError>>, SpannedError>,
}

impl<'source> Hash<'source> {
    pub fn new(
        token: Token<'source>,
        kv_pairs: Result<Vec<Result<ExprPairs<'source>, SpannedError>>, SpannedError>,
    ) -> Self {
        Self { token, kv_pairs }
    }
}

#[derive(Debug, PartialEq)]
pub struct Index<'source> {
    token: Token<'source>,
    object: Box<Expression<'source>>,
    index: Box<ExprResult<'source>>,
}

impl<'source> Index<'source> {
    pub fn new(
        token: Token<'source>,
        object: Expression<'source>,
        index: ExprResult<'source>,
    ) -> Self {
        Self {
            token,
            object: Box::new(object),
            index: Box::new(index),
        }
    }
}

macro_rules! expr_impls {
    ($($expr:tt),+) => {$(
        impl<'source> From<$expr<'source>> for Node<'source> {
            fn from(expr: $expr<'source>) -> Self {
                let expression: Expression = expr.into();
                expression.into()
            }
        }

        impl<'source> From<$expr<'source>> for Expression<'source> {
            fn from(expr: $expr<'source>) -> Self {
                Expression::$expr(expr)
            }
        }
    )+}
}

expr_impls!(
    Identifier,
    Primative,
    StringLiteral,
    Prefix,
    Infix,
    If,
    Function,
    Call,
    Array,
    Hash,
    Index
);
