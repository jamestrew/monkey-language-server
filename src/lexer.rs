use std::ops::Range;

use logos::Logos;

use crate::diagnostics::{MonkeyError, SpannedError};
use crate::parser::TokenProvider;
use crate::types::{Position, Spanned};

#[derive(Logos, Debug, PartialEq, Clone, Copy)]
#[logos(skip r"[ \t\f]")]
pub enum TokenKind {
    #[token("=")]
    Assign,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Asterisk,
    #[token("/")]
    ForwardSlash,
    #[token("!")]
    Bang,

    #[token("==")]
    Equal,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LT,
    #[token(">")]
    GT,

    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[token(":")]
    Colon,

    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,

    // keywords
    #[token("let")]
    Let,
    #[token("fn")]
    Function,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("true")]
    True,
    #[token("false")]
    False,
    #[token("return")]
    Return,
    #[token("nil")]
    Nil,

    #[regex("[a-zA-Z][a-zA-Z0-9_]*")]
    Identifier,
    #[regex("[0-9]+")]
    Int,
    #[regex(r#""(?:[^"]|\\")*""#)]
    Str,

    #[regex("\n|\r\n")]
    NewLine,
}

impl AsRef<TokenKind> for TokenKind {
    fn as_ref(&self) -> &TokenKind {
        self
    }
}

impl std::fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenKind::Assign => write!(f, "="),
            TokenKind::Plus => write!(f, "+"),
            TokenKind::Minus => write!(f, "-"),
            TokenKind::Asterisk => write!(f, "*"),
            TokenKind::ForwardSlash => write!(f, "/"),
            TokenKind::Bang => write!(f, "!"),
            TokenKind::Equal => write!(f, "=="),
            TokenKind::NotEqual => write!(f, "!="),
            TokenKind::LT => write!(f, "<"),
            TokenKind::GT => write!(f, ">"),
            TokenKind::Comma => write!(f, ","),
            TokenKind::Semicolon => write!(f, ";"),
            TokenKind::Colon => write!(f, ":"),
            TokenKind::LParen => write!(f, "("),
            TokenKind::RParen => write!(f, ")"),
            TokenKind::LBrace => write!(f, "{{"),
            TokenKind::RBrace => write!(f, "}}"),
            TokenKind::LBracket => write!(f, "["),
            TokenKind::RBracket => write!(f, "]"),
            TokenKind::Let => write!(f, "let"),
            TokenKind::Function => write!(f, "fn"),
            TokenKind::If => write!(f, "if"),
            TokenKind::Else => write!(f, "else"),
            TokenKind::True => write!(f, "true"),
            TokenKind::False => write!(f, "false"),
            TokenKind::Return => write!(f, "return"),
            TokenKind::Nil => write!(f, "nil"),
            TokenKind::Identifier => write!(f, "<identifier>"),
            TokenKind::Int => write!(f, "<int>"),
            TokenKind::Str => write!(f, "<str>"),
            TokenKind::NewLine => write!(f, r"\n"),
        }
    }
}

#[derive(PartialEq)]
pub struct _Token<'source> {
    pub kind: TokenKind,
    pub slice: &'source str,
}

impl<'source> _Token<'source> {
    #[cfg(test)]
    pub fn new(kind: TokenKind, slice: &'source str) -> _Token<'source> {
        Self { kind, slice }
    }

    pub fn from_lexer(lexer: &Lexer<'source>, kind: TokenKind) -> Token<'source> {
        let slice = lexer.current_slice();
        lexer.new_spanned_item(Self { kind, slice })
    }
}

pub type Token<'source> = Spanned<_Token<'source>>;
pub type TokenResult<'source> = Result<Token<'source>, SpannedError>;

pub fn token_result_span<T>(token_res: &TokenResult<'_>, data: T) -> Spanned<T> {
    match token_res {
        Ok(token) => token.map(data),
        Err(err) => err.map(data),
    }
}

pub struct Lexer<'source> {
    lexer: logos::Lexer<'source, TokenKind>,
    row: usize,
    last_newline_pos: usize,
}

impl<'source> Lexer<'source> {
    pub fn new(source: &'source str) -> Self {
        let lexer = TokenKind::lexer(source);

        Self {
            lexer,
            row: 0,
            last_newline_pos: 0,
        }
    }

    fn new_spanned_item<T>(&self, data: T) -> Spanned<T> {
        let span = self.lexer.span();
        let col_start = span.start - self.last_newline_pos;
        let col_end = span.end - self.last_newline_pos;

        let start = Position::new(self.row, col_start);
        let end = Position::new(self.row, col_end);

        Spanned::new(start, end, self.current_span(), data)
    }

    pub fn current_span(&self) -> Range<usize> {
        self.lexer.span().clone()
    }

    fn current_slice(&self) -> &'source str {
        self.lexer.slice()
    }
}

impl<'source> std::fmt::Debug for Token<'source> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Token({:#?}, {:#?}, {})",
            self.kind,
            self.slice,
            self.pos_rng_str()
        )
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = TokenResult<'source>;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some(token_res) = self.lexer.next() {
            match token_res {
                Ok(TokenKind::NewLine) => {
                    self.row += 1;
                    self.last_newline_pos = self.lexer.span().end;
                    continue;
                }
                Ok(token) => return Some(Ok(_Token::from_lexer(self, token))),
                Err(_) => {
                    let err = self
                        .new_spanned_item(MonkeyError::UnexpectedToken(self.lexer.slice().into()));
                    return Some(Err(err));
                }
            }
        }
        None
    }
}

impl<'source> TokenProvider<'source> for Lexer<'source> {
    fn next(&mut self) -> Option<TokenResult<'source>> {
        Iterator::next(self)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    enum T<'source> {
        Token(Token<'source>),
        Error(SpannedError),
    }

    impl<'source> std::fmt::Debug for T<'source> {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            match self {
                Self::Token(t) => write!(f, "{:?}", t),
                Self::Error(err) => write!(f, "{:?}", err),
            }
        }
    }

    fn debug_print(token_results: Vec<Result<Token<'_>, SpannedError>>) -> Vec<T> {
        let mut ret = Vec::new();
        for result in token_results {
            match result {
                Ok(token) => ret.push(T::Token(token)),
                Err(err) => ret.push(T::Error(err)),
            }
        }
        ret
    }

    macro_rules! snapshot {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let lexer = Lexer::new($input);
                let tokens: Vec<_> = lexer.collect();

                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_debug_snapshot!(debug_print(tokens));
                })
            }
        };
    }

    snapshot!(check_span, "+\n+\n   +");
    snapshot!(simple_error, "@");
    snapshot!(simple_symbol, "+");
    snapshot!(more_symbols_1, "= + - *  / !");
    snapshot!(more_symbols_2, "==\n!=\n\t<>,;:");
    snapshot!(more_symbols_3, "(){}[]");
    snapshot!(keywords, "let\nfn\nif\nelse\ntrue\nfalse\nreturn nil");
    snapshot!(identifiers, "x\nabc");
    snapshot!(ints, "123");
    snapshot!(str, "\"hello world 123\"");

    snapshot!(assignments1, "let five = 5;");
    snapshot!(assignments2, "let ten = 10;");
    snapshot!(
        assignments3,
        r"let add = fn(x, y) {
            x + y;
        };"
    );
    snapshot!(assignments4, "let result = add(five, ten);");

    snapshot!(operators, "!-/*5;");
    snapshot!(equality1, "5 < 10 > 5;");
    snapshot!(equality2, "10 == 10;");
    snapshot!(equality3, "10 != 9;");
    snapshot!(
        conditional,
        r"if (5 < 10) {
            return true;
        } else {
            return false;
        }"
    );

    snapshot!(arrays, "[1, 2]");
    snapshot!(dictionary, r#"{ "foo": "bar" };"#);
}
