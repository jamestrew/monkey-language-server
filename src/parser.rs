#![allow(dead_code)]
use crate::ast::{
    ExprResult, Identifier, Let, Node, Primative, Program, Return, StmtResult, StringLiteral,
};
use crate::errors::{MonkeyError, SpannedError};
use crate::lexer::{token_result_span, Lexer, Token, TokenKind, TokenResult};
use crate::types::Spanned;

pub struct Parser<'source> {
    lexer: Lexer<'source>,
    curr_token: Option<TokenResult<'source>>,
    peek_token: Option<TokenResult<'source>>,
    last_span: Spanned<()>,
}

impl<'source> Parser<'source> {
    pub fn new(source: &'source str) -> Parser {
        let mut lexer = Lexer::new(source);
        let curr_token = lexer.next();
        let peek_token = lexer.next();

        Self {
            lexer,
            curr_token,
            peek_token,
            last_span: Spanned::default(),
        }
    }

    pub fn parse_program(&mut self) -> Program<'source> {
        let mut nodes = Vec::new();

        while self.curr_token.is_some() {
            match self.parse_statement() {
                Ok(stmt) => nodes.push(Node::Statement(stmt)),
                Err(err) => nodes.push(Node::Error(err)),
            }
        }

        Program { nodes }
    }

    fn next_token(&mut self) -> TokenResult<'source> {
        let ret = match self.curr_token.take() {
            Some(token_res) => {
                self.last_span = token_result_span(&token_res, ());
                token_res
            }
            None => Err(self.last_span.map(MonkeyError::UnexpectedEof)),
        };

        self.curr_token = self.peek_token.take();
        self.peek_token = self.lexer.next();
        ret
    }

    fn current_token_is<T: AsRef<TokenKind>>(
        &mut self,
        match_token: T,
    ) -> Result<bool, SpannedError> {
        match self.curr_token.as_ref() {
            Some(Ok(token)) => Ok(token.kind == *match_token.as_ref()),
            Some(Err(err)) => Err(err.clone_inner()),
            None => Ok(false),
        }
    }

    fn peek_token_is<T: AsRef<TokenKind>>(&mut self, match_token: T) -> Result<bool, SpannedError> {
        match &self.curr_token {
            Some(Ok(ref token)) => Ok(*match_token.as_ref() == token.kind),
            Some(Err(err)) => Err(err.clone_inner()),
            None => Ok(false),
        }
    }

    fn expect_current<T: AsRef<TokenKind>>(&mut self, expect_token: T) -> TokenResult<'source> {
        match &self.curr_token {
            Some(Ok(token)) => {
                if token.kind == *expect_token.as_ref() {
                    self.next_token()
                } else {
                    Err(token.map(MonkeyError::ExpectedTokenNotFound(
                        expect_token.as_ref().to_string(),
                    )))
                }
            }
            Some(Err(err)) => Err(err.clone_inner()),
            None => Err(self.last_span.map(MonkeyError::UnexpectedEof)),
        }
    }

    fn expect_peek<T: AsRef<TokenKind>>(&mut self, expect_token: T) -> TokenResult<'source> {
        match &self.peek_token {
            Some(Ok(ref token)) => {
                if token.kind == *expect_token.as_ref() {
                    self.next_token()
                } else {
                    Err(token.map(MonkeyError::ExpectedTokenNotFound(token.slice.into())))
                }
            }
            Some(Err(err)) => Err(err.clone_inner()),
            None => {
                // token_result_span(&self.curr_token.expect("have valid curr_token"), ());
                todo!();
            }
        }
    }

    fn eat_semicolons(&mut self) -> Result<(), SpannedError> {
        while self.peek_token_is(TokenKind::Semicolon)? {
            self.next_token()?;
        }
        Ok(())
    }

    fn parse_statement(&mut self) -> StmtResult<'source> {
        let token = self.next_token()?;

        let result = match token.kind {
            TokenKind::Let => self.parse_let_statement(token),
            TokenKind::Return => self.parse_return_statement(token),
            _ => self
                .parse_expression_statement(token)
                .map(|expr| expr.into()),
        };

        if result.is_err() {
            self.eat_semicolons()?;
        }
        result
    }

    fn parse_let_statement(&mut self, token: Token<'source>) -> StmtResult<'source> {
        let ident_token = self.expect_current(TokenKind::Identifier)?;
        self.expect_current(TokenKind::Assign)?;
        let value_token = self.next_token()?;

        let let_stmt = Let::new(
            token,
            Identifier::from(ident_token),
            self.parse_expression_statement(value_token)?,
        )
        .into();
        self.eat_semicolons()?;
        Ok(let_stmt)
    }

    fn parse_return_statement(&mut self, token: Token<'source>) -> StmtResult<'source> {
        let value;
        if self.current_token_is(TokenKind::Semicolon)? {
            value = None;
        } else {
            let value_token = self.next_token()?;
            value = Some(self.parse_expression_statement(value_token)?);
        }
        let return_stmt = Return::new(token, value).into();
        self.eat_semicolons()?;
        Ok(return_stmt)
    }

    fn parse_expression_statement(&mut self, token: Token<'source>) -> ExprResult<'source> {
        match &token.kind {
            TokenKind::Identifier => Ok(Identifier::from(token).into()),
            TokenKind::Int | TokenKind::True | TokenKind::False | TokenKind::Nil => {
                Ok(Primative::from(token).into())
            }
            TokenKind::Str => Ok(StringLiteral::from(token).into()),
            _ => Err(token.map(MonkeyError::UnexpectedToken(token.slice.into()))),
        }
    }
}

#[cfg(test)]
mod test {
    use std::ops::Range;

    use super::*;
    use crate::lexer::*;
    use crate::types::*;

    pub fn debug_new(
        start: Position,
        end: Position,
        span: Range<usize>,
        kind: TokenKind,
        slice: &'_ str,
    ) -> Token<'_> {
        Spanned::new(start, end, span, _Token::new(kind, slice))
    }

    #[test]
    fn new_parser() {
        let source = "let a = 123;";
        let parser = Parser::new(source);

        let first = debug_new(
            Position::new(0, 0),
            Position::new(0, 3),
            0..3,
            TokenKind::Let,
            "let",
        );
        let second = debug_new(
            Position::new(0, 4),
            Position::new(0, 5),
            4..5,
            TokenKind::Identifier,
            "a",
        );

        assert_eq!(parser.curr_token, Some(Ok(first)));
        assert_eq!(parser.peek_token, Some(Ok(second)));
    }

    macro_rules! debug_snapshot {
        ($name:tt, $input:expr) => {
            #[test]
            fn $name() {
                let stmts = Parser::new($input).parse_program();
                insta::with_settings!({
                    description => $input,
                }, {
                    insta::assert_debug_snapshot!(stmts);
                })
            }
        };
    }

    debug_snapshot!(lexer_erro, "@");
    debug_snapshot!(bool_expr, "true");

    // maybe these are warnings? "Expression value is unused" <- python
    debug_snapshot!(int_expr, "123");
    debug_snapshot!(nil_expr, "nil");
    debug_snapshot!(string_literal, "\"hello world\"");
    debug_snapshot!(identifier, "x");

    debug_snapshot!(let_statement_1, "let x = 1;");
    debug_snapshot!(let_statement_2, "let x = 2;;;");
    debug_snapshot!(let_statement_broken_1, "let x = ;");
    debug_snapshot!(let_statement_broken_2, "let x;");
    debug_snapshot!(let_statement_broken_3, "let;");
    // debug_snapshot!(let_statement_broken_4, "le a = 1;");

    debug_snapshot!(return_happy_1, "return 1;");
    debug_snapshot!(return_happy_2, "return nil;");
    debug_snapshot!(return_happy_3, "return;");
}
