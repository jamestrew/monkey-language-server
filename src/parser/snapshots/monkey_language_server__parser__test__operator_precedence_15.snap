---
source: src/parser.rs
description: 3 > 5 == false
expression: stmts
---
[
    Infix {
        token: Token(Equal, "==", (0,6)->(0,8)),
        left: Infix {
            token: Token(GT, ">", (0,2)->(0,3)),
            left: Int {
                token: Token(Int, "3", (0,0)->(0,1)),
                value: 3,
            },
            right: Int {
                token: Token(Int, "5", (0,4)->(0,5)),
                value: 5,
            },
        },
        right: Bool {
            token: Token(False, "false", (0,9)->(0,14)),
            value: false,
        },
    },
]
